File: sample-videochat-webrtc/sample-videochat-webrtc/Components/CallPermisions/CallPermissions.m
//
//  CallPermissions.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "CallPermissions.h"

@implementation CallPermissions

+ (void)checkPermissionsWithConferenceType:(QBRTCConferenceType)conferenceType
                                completion:(GrantedPermissionBlock)completion {
    
#if TARGET_OS_SIMULATOR
    completion(YES);
    return;
#else
    
    [self requestPermissionToMicrophoneWithCompletion:^(BOOL granted) {
        
        if (granted) {
            
            switch (conferenceType) {
                    
                case QBRTCConferenceTypeAudio:
                    
                    if (completion) {
                        
                        completion(granted);
                    }
                    
                    break;
                    
                case QBRTCConferenceTypeVideo: {
                    
                    [self requestPermissionToCameraWithCompletion:^(BOOL videoGranted) {
                        
                        if (!videoGranted) {
                            
                            // showing error alert with a suggestion
                            // to go to the settings
                            [self showAlertWithTitle:NSLocalizedString(@"Camera error", nil)
                                             message:NSLocalizedString(@"The app doesn't have access to the camera, please go to settings and enable it.", nil)];
                        }
                        
                        if (completion) {
                            
                            completion(videoGranted);
                        }
                    }];
                    
                    break;
                }
            }
        }
        else {
            // showing error alert with a suggestion
            // to go to the settings
            [self showAlertWithTitle:NSLocalizedString(@"Microphone error", nil)
                             message:NSLocalizedString(@"The app doesn't have access to the microphone, please go to settings and enable it.", nil)];
            
            if (completion) {
                
                completion(granted);
            }
        }
    }];
#endif
}

+ (void)requestPermissionToMicrophoneWithCompletion:(GrantedPermissionBlock)completion {
    
    [[AVAudioSession sharedInstance] requestRecordPermission:^(BOOL granted) {
        if (completion) {
            dispatch_async(dispatch_get_main_queue(), ^{completion(granted);});
        }
    }];
}

+ (void)requestPermissionToCameraWithCompletion:(GrantedPermissionBlock)completion {
    
    NSString *mediaType = AVMediaTypeVideo;
    AVAuthorizationStatus authStatus = [AVCaptureDevice authorizationStatusForMediaType:mediaType];
    switch (authStatus) {
            
        case AVAuthorizationStatusNotDetermined: {
            
            [AVCaptureDevice requestAccessForMediaType:mediaType completionHandler:^(BOOL granted) {
                
                if (completion) {
                    dispatch_async(dispatch_get_main_queue(), ^{completion(granted);});
                }
            }];
            
            break;
        }
            
        case AVAuthorizationStatusRestricted:
        case AVAuthorizationStatusDenied:
            
            if (completion) {
                completion(NO);
            }
            
            break;
            
        case AVAuthorizationStatusAuthorized:
            
            if (completion) {
                completion(YES);
            }
            
            break;
    }
}

#pragma mark - Helpers

+ (void)showAlertWithTitle:(NSString *)title message:(NSString *)message {
    
    UIAlertController *alertController =
    [UIAlertController alertControllerWithTitle:title
                                        message:message
                                 preferredStyle:UIAlertControllerStyleAlert];
    
    [alertController addAction:
     [UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel", nil)
                              style:UIAlertActionStyleCancel
                            handler:^(UIAlertAction * _Nonnull __unused action){/*Empty action*/}]
     ];
    
    [alertController addAction:
     [UIAlertAction actionWithTitle:NSLocalizedString(@"Settings", nil)
                              style:UIAlertActionStyleDefault
                            handler:^(UIAlertAction * _Nonnull __unused action) {
                                
                                NSURL *settingsUrl = [NSURL URLWithString:UIApplicationOpenSettingsURLString];
                                [[UIApplication sharedApplication] openURL:settingsUrl options:@{} completionHandler:nil];
                            }]
     ];
    
    [[UIApplication sharedApplication].keyWindow.rootViewController presentViewController:alertController
                                                                                 animated:YES
                                                                               completion:nil];
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/AVPermisions/CallPermissions.swift
//
//  CallPermissions.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 12/10/18.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import Foundation
import QuickbloxWebRTC

struct CallErrorConstant {
    static let cameraErrorTitle = NSLocalizedString("Camera error", comment: "")
    static let cameraErrorMessage = NSLocalizedString("The app doesn't have access to the camera, please go to settings and enable it.", comment: "")
    static let microphoneErrorTitle = NSLocalizedString("Microphone error", comment: "")
    static let microphoneErrorMessage = NSLocalizedString("The app doesn't have access to the microphone, please go to settings and enable it.", comment: "")
    static let alertCancelAction = NSLocalizedString("Cancel", comment: "")
    static let alertSettingsAction = NSLocalizedString("Settings", comment: "")
}

class CallPermissions {
    //MARK: - Class Methods
    class func check(with conferenceType: QBRTCConferenceType, completion: @escaping (_ granted: Bool) -> Void ) {
        
        #if targetEnvironment(simulator)
        completion(true)
        return
        #endif
        
        self.requestPermissionToMicrophone(withCompletion: { granted in
            guard granted == true else {
                showAlert(withTitle: CallErrorConstant.microphoneErrorTitle,
                          message: CallErrorConstant.microphoneErrorMessage)
                completion(granted)
                return
            }
            switch conferenceType {
            case .audio: completion(granted)
            case .video:
                requestPermissionToCamera(withCompletion: { videoGranted in
                    if videoGranted == false {
                        showAlert(withTitle: CallErrorConstant.cameraErrorTitle,
                                  message: CallErrorConstant.cameraErrorMessage)
                    }
                    completion(videoGranted)
                })
            }
        })
    }
    
    class func requestPermissionToMicrophone(withCompletion completion: @escaping (_ granted: Bool) -> Void ) {
        AVAudioSession.sharedInstance().requestRecordPermission({ granted in
            DispatchQueue.main.async(execute: {
                completion(granted)
            })
        })
    }
    
    class func requestPermissionToCamera(withCompletion completion: @escaping (_ granted: Bool) -> Void ) {
        let mediaType = AVMediaType.video
        let authStatus = AVCaptureDevice.authorizationStatus(for: mediaType)
        switch authStatus {
        case .notDetermined:
            AVCaptureDevice.requestAccess(for: mediaType, completionHandler: { granted in
                DispatchQueue.main.async(execute: {
                    completion(granted)
                })
            })
        case .restricted, .denied:
            completion(false)
        case .authorized:
            completion(true)
        }
    }
    
    // MARK: - Helpers
    // showing error alert with a suggestion
    // to go to the settings
    class func showAlert(withTitle title: String?, message: String?) {
        let alertController = UIAlertController(title: title,
                                                message: message,
                                                preferredStyle: .alert)
        alertController.addAction(UIAlertAction(title: CallErrorConstant.alertCancelAction,
                                                style: .cancel))
        alertController.addAction(UIAlertAction(title: CallErrorConstant.alertSettingsAction,
                                                style: .default,
                                                handler: { action in
                                                    if let url = URL(string: UIApplication.openSettingsURLString) {
                                                        UIApplication.shared.open(url, options: [:])
                                                    }
        }))
        UIApplication.shared.keyWindow?.rootViewController?.present(alertController, animated: true)
    }
}
